nn/conformer.py

```diff
@@ -0,0 +1,217 @@
+    self.pool_sizes = pool_sizes
+    self.activation = activation
+
+    self.conv_layers = nn.ModuleList()
+    assert len(filter_sizes) == len(channel_sizes)
+    for filter_size, channel_size in zip(filter_sizes, channel_sizes):
+      self.conv_layers.append(
+        nn.Conv(filter_size=filter_size, n_out=channel_size, padding=padding))
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    x = nn.split_dims(inp, axis='F', dims=(-1, 1))
+    for i, conv_layer in enumerate(self.conv_layers):
+      x = conv_layer(x)
+      x = self.activation(x)
+      if self.pool_sizes and i < len(self.pool_sizes):
+        x = nn.pool(x, pool_size=self.pool_sizes[i], padding='same', mode='max')
+      if self.dropout:
+        x = nn.dropout(x, dropout=self.dropout)
+    out = nn.merge_dims(x, axes='static')
+    return out
+
+
+class ConformerEncoderLayer(nn.Module):
+  """
+  Represents a conformer block
+  """
+
+  def __init__(
+        self, conv_kernel_size: int, activation_ff: Callable[[nn.LayerRef], nn.LayerRef], dim_ff: int, dropout: float,
+        att_dropout: float, enc_key_dim: int, num_heads: int, batch_norm_opts: Optional[Dict[str, Any]] = None):
+    """
+    :param conv_kernel_size:
+    :param activation_ff:
+    :param ff_dim:
+    :param dropout:
+    :param att_dropout:
+    :param enc_key_dim:
+    :param num_heads:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+
+    self.ffn1 = _PositionwiseFeedForward(
+      out_dim=enc_key_dim, dim_ff=dim_ff, dropout=dropout, activation=activation_ff)
+
+    self.ffn2 = _PositionwiseFeedForward(
+      out_dim=enc_key_dim, dim_ff=dim_ff, dropout=dropout, activation=activation_ff)
+
+    self.conv_block = _ConformerConvBlock(
+      out_dim=enc_key_dim, kernel_size=conv_kernel_size, batch_norm_opts=batch_norm_opts)
+
+    self.self_att = MultiheadAttention(enc_key_dim, num_heads, dropout=att_dropout)  # TODO: to be implemented
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    # FFN
+    x_ffn1_ln = nn.layer_norm(inp)
+    x_ffn1 = self.ffn1(x_ffn1_ln)
+    x_ffn1_out = 0.5 * nn.dropout(x_ffn1, dropout=self.dropout) + inp
+
+    # MHSA
+    x_mhsa_ln = nn.layer_norm(x_ffn1_out)
+    x_mhsa = self.self_att(x_mhsa_ln)
+    x_mhsa_out = x_mhsa + x_ffn1_out
+
+    # Conv
+    x_conv_ln = nn.layer_norm(x_mhsa_out)
+    x_conv = self.conv_block(x_conv_ln)
+    x_conv_out = nn.dropout(x_conv, dropout=self.dropout) + x_mhsa_out
+
+    # FFN
+    x_ffn2_ln = nn.layer_norm(x_conv_out)
+    x_ffn2 = self.ffn2(x_ffn2_ln)
+    x_ffn2_out = 0.5 * nn.dropout(x_ffn2, dropout=self.dropout) + x_conv_out
+
+    # last LN layer
+    return nn.layer_norm(x_ffn2_out)
+
+
+class ConformerEncoder(nn.Module):
+  """
+  Represents Conformer encoder architecture
+  """
+
+  def __init__(
+        self, encoder_layer: nn.Module, num_blocks: int, conv_kernel_size: int = 32,
+        activation_ff: Callable[[nn.LayerRef], nn.LayerRef] = nn.swish, dim_ff: int = 512, dropout: float = 0.1,
```

Comment:
Here you have also dropout defaults, and also att_dropout.
I'm just curious. Is this from the paper? Or ESPNet? Or both? Are these defaults good?
Also for dimension, num heads.
Does PyTorch Transformer has such defaults?

Reply:
I think this was just a good starting point when I started doing experiments with conformer. But we can follow the parmas of ESPNet best conformer config as here: https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_asr_conformer7_n_fft512_hop_length256.yaml

Reply:
But I wonder whether we should set any defaults at all. Or which defaults. It seems kind of arbitrary.
Although I can see that some defaults make sense.
Is this the same as in the original paper?

Why is there no default for `encoder_layer` and `num_blocks`?
Btw, instead of `num_blocks`, maybe `num_layers` is better?
(Like [PyTorch Transformer](https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html#torch.nn.Transformer).)

Reply:
> Is this the same as in the original paper?
Not everything. The number of layers is 16 in the original paper. I think the feed forward dimension is not mentioned there. I would suggest to use ESPNet parameters for now. I think i also used the same parameters for the conformer-based hybrid model.

Reply:
Ok. But note, as usual, for default values, once we settled on these, we cannot easily change them anymore (unless we introduce sth like behavior version here as well...). So they should be reasonable. Nothing experimental.

Reply:
> Why is there no default for encoder_layer and num_blocks?

Similar to [PyTorch TransformerEncoder](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html)

Reply:
But there the code does not create the encoder layer at all, so all the other options are also not needed. Why don't you make it consistent to that anyway?